solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1702,8 +1702,13 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
       }
       DocCollection coll = zkStateReader.getCollection(collection);
       if (forcePublish || sendToOverseer(coll, coreNodeName)) {
+        //for PRS, this only updates state.json but not the actual PRS entries
         overseerJobQueue.offer(Utils.toJSON(m));
-      } else {
+      }
+
+      //extra handling for PRS, we need to write the PRS entries from this node directly,
+      //as overseer does not and should not handle those entries
+      if (coll != null && coll.isPerReplicaState() && coreNodeName != null) {
```

Comment:
@patsonluk here we are might update state.json through overseer above. Likely those will be shard's state or other params. But we need to make sure race condition between these two operations(overseer state.json update vs direct prs) should not cause any other issue.

Reply:
Race conditions that concern me so far are the state.json vs state.json or PRS entries vs PRS entries. Do you think the ordering of update on state.json vs PRS entries would cause any issues? :)

Reply:
The state.json maintains collection structure and prs state maintains replica-status. So better to keep this semantics. 

Also, good to see if we don't change this then what fails!

Reply:
agree with @patsonluk here. This is unlikely to cause any inconsistency.

Reply:
@hiteshk25 As discussed before, your concern is that this is a "change of behavior", as before this change, the `state.json` update AND the PRS entries update (if `forcePublish || sendToOverseer(coll, coreNodeName) evaluates to `true`) happen at the same piece of code (`ReplicaMutator#updateState`) and in predictable execution order - state.json update first then PRS entries. 

With the code change here, it's very likely that PRS entries update would happen BEFORE the state.json update, and the ordering is not deterministic.

Hence the only weaker guarantee we have with this new change, is that "eventually" the PRS and state.json will both be updated.

For the several places that watch for replica state changes, they seem to be okay with this weaker guarantee, such as [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1881) in Solr java code and [here](https://github.com/cowpaths/mn/blob/master/projects/fullstory/go/src/fs/services/solrman/internal/solrmansvc/move.go#L112) in our go code

Can you check if any code u know would have issue with such weaker guarantee?

I agree that ideally we want to keep the deterministic execute order as before, however it could be a bit challenging to implement it now as `waitForState` probably wouldn't work for PRS collection replica entry update as it will not be able to see the change in PRS state after sending it to OS as we now move it away from the overseer handling.  😓

Reply:
As discussed with @hiteshk25, going to document the findings here:

Our concern for this code is that if state.json update and PRS update are not in the correct order. Then we could risk the PRS update being dropped by our go solr code (a long description added to very bottom)

However, if `state.json` already has the replica in the json content, then it's relatively safer. Hence, we are going to assess operations that we care, and see if `state.json` already knows about the replica:

1. Split/move shard - which eventually call ADDREPLICA. It is safe as it first call `SliceMutator#addReplica` which [translate to a ZkWriteCommand that updates the `state.json`](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/overseer/SliceMutator.java#L101). Then it will [wait](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/api/collections/AddReplicaCmd.java#L265) to see the replica in place first before proceeding. Later on the `ZkController#preRegister` will have `forcePublish` as `true`
2. Create collection - Also safe as `CreateCollectionCmd` [updates the `state.json`](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L282) before the cores are created (which later call `ZkController#preRegister` which triggers this code

Reply:
**This comment is below solely for documentation purpose. NOT that important for current situation**

it could be a bit problematic if PRS entry is updated BEFORE the corresponding replica is added to `state.json`. 

Normally, when the replica details is added to state.json first. it would trigger the "data notification" with the state.json as bytes, such byte array will be parsed and fill up the `collection` struct (which has collection -> shard -> replica info). For PRS collection, the replica info will then be ["replace" by](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L622) whatever is in the old/existing `collection` struct, last observed (if existed) replica state triggered by child watch. Then when the PRS entries are updated, it will trigger the "child watch" with the PRS entry string, it will then [update](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L278) the replica info in the `collection` struct.

Now consider the cases that IF PRS entry is updated BEFORE the corresponding replica is added: The child watch routine will now not do anything as in the [code](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L278) linked earlier, if such replica does not existing in the `collection` struct yet, the update will be skipped.  Then when the state.json is updated, IF somehow the state.json has matching replica details as the PRS entry, then *it will be okay*. It will find no previous replica details to replace with and just use the details in  the incoming state.json. However, if the state.json does not contain the correct replica details (active/leadership, and possible more details in the future), then the replica details in `collection` struct will be wrong.

This is probably not super major now as in `ZkController#publish` the `Replica.State state` (used to publish to OS) and `DocCollection#getPerReplicaStates` (for direct PRS entries writes) are supposed to carry the same replica details.